codec/api.go

```diff
@@ -1,48 +1,124 @@
 package codec
 
 import (
-	"github.com/ipld/go-ipld-prime"
+	"io"
+
+	"github.com/ipld/go-ipld-prime/datamodel"
 )
 
-// Encoder is defined in the root ipld package; this alias is just for documentation and discoverability.
+// The following two types define the two directions of transform that a codec can be expected to perform:
+// from Node to serial stream (aka "encoding", sometimes also described as "marshalling"),
+// and from serial stream to Node (via a NodeAssembler) (aka "decoding", sometimes also described as "unmarshalling").
+//
+// You'll find a couple of implementations matching this shape in subpackages of 'codec'.
+// (These are the handful of encoders and decoders we ship as "batteries included".)
+// Other encoder and decoder implementations can be found in other repositories/modules.
+// It should also be easy to implement encodecs and decoders of your own!
+//
+// Encoder and Decoder functions can be used on their own, but are also often used via the `ipld/linking.LinkSystem` construction,
+// which handles all the other related operations necessary for a content-addressed storage system at once.
+//
+// Encoder and Decoder functions can be registered in the multicodec table in the `ipld/multicodec` package
+// if they're providing functionality that matches the expectations for a multicodec identifier.
+// This table will be used by some common EncoderChooser and DecoderChooser implementations
+// (namely, the ones in LinkSystems produced by the `linking/cid` package).
+// It's not strictly necessary to register functions there, though; you can also just use them directly.
 //
-// Encoder functions can be composed into an ipld.LinkSystem to provide
-// a "one stop shop" API for handling content addressable storage.
-// Encoder functions can also be used directly if you want to handle serial data streams.
+// There are furthermore several conventions that codec packages are recommended to follow, but are only conventions:
 //
-// Most codec packages will have a ReusableEncoder type
-// (which contains any working memory needed by the encoder implementation,
-// as well as any configuration options),
-// and that type will have an Encode function matching this interface.
+// Most codec packages should have a ReusableEncoder and ResuableDecoder type,
```

Comment:
do these `Reusable*` things exist or are they just an idea? I only see mentions in this doc in this PR and master only has it in here and in a comment in codec/dagjson2/doc.go.

Reply:
Unfortunately, this is speculative and wishful, yes.

As it says "should", I'm keeping it.